scipio/src/sys/uring.rs

```diff
@@ -489,6 +489,75 @@ impl UringCommon for PollRing {
     }
 }
 
+impl InnerSource {
+    pub(crate) fn update_source_type(&mut self, source_type: SourceType) -> SourceType {
+        std::mem::replace(&mut self.source_type, source_type)
+    }
+}
+
+#[allow(clippy::mut_from_ref)]
+// This is similar to the interior mutability pattern but clippy doesn't like it.
+// That's because unlike the usual refcell patterns, we return a mutable reference
+// and the user is free to do whatever it wants with it (as opposed to a scoped RefMut).
+//
+// We don't have too much of a choice because the Source accompanies a request in
+// the io_uring and we'd be dealing with raw pointers anyway. We want to make this as
+// safe as possible but will always have to rely on those weird tricks.
+//
+// This should only really be used within the io_uring code
+fn mut_source(source: &Rc<UnsafeCell<InnerSource>>) -> &mut InnerSource {
+    unsafe { &mut *source.get() }
+}
+
+impl Source {
+    #[allow(clippy::mut_from_ref)]
+    fn inner(&self) -> &mut InnerSource {
+        mut_source(&self.inner)
+    }
+
+    fn update_reactor_info(&self, id: u64, queue: ReactorQueue) {
+        self.inner().id = Some(id);
+        self.inner().queue = Some(queue);
+    }
+
+    fn consume_id(&self) -> Option<u64> {
+        self.inner().id.take()
+    }
+
+    fn latency_req(&self) -> Latency {
+        self.inner().io_requirements.latency_req
+    }
+
+    fn source_type(&self) -> &SourceType {
+        &self.inner().source_type
+    }
+
+    fn update_source_type(&mut self, source_type: SourceType) -> SourceType {
+        self.inner().update_source_type(source_type)
+    }
+
+    pub(crate) fn extract_source_type(&mut self) -> SourceType {
+        self.inner().update_source_type(SourceType::Invalid)
+    }
+
+    pub(crate) fn wakers(&self) -> &RefCell<Wakers> {
+        &self.inner().wakers
+    }
```

Comment:
Your commit message states that:
```
The functions that return references are all internal now, and the
ones that are pub(crate) are all returning copies of parts of the source
the external entities need to manipulate (type, raw fd, etc)
```
But this function is public _and_ returns a reference.

Reply:
right, but that's a reference to a RefCell, meaning access will happen through the usual borrows

It is still bad wording so I will change.

Reply:
I went one step further and actually changed the code so that the reference is not returned.
It costs us an extra borrow_mut(), but it's better layering and better isolation so I think it is worth it

Reply:
@HippoBaro please take a look at the end result